### PR TITLE
StaticResource only matches folder-like URL prefix

### LINKED ISSUE #5250: router.add_static is greedy
🐞 **Describe the bug**

The first argument for `add_static` matches on any path that starts with that string. For instance, if you specify "h", then `add_static` will match on "h" or "help" or "hoover".

💡 **To Reproduce**

Run this code:

```
from aiohttp import web

async def handle(request):
    name = request.match_info.get('name', "Anonymous")
    text = "Hello, " + name
    return web.Response(text=text)

app = web.Application()

router = app.router

router.add_static("/h", "./")

router.add_routes([web.get('/', handle),
                web.get('/{name}', handle)])

if __name__ == '__main__':
    web.run_app(app)
```

Add a text file in the current working directory named `test.txt` and put some text in there.

Then navigate to the following URLs to demonstrate the issue:

* http://0.0.0.0:8080/
* http://0.0.0.0:8080/groot
* http://0.0.0.0:8080/h/test.txt
* http://0.0.0.0:8080/hoover

The last link unexpectedly returns a 404.

💡 **Expected behavior**

I would have expected http://0.0.0.0:8080/hoover to have similar behavior to http://0.0.0.0:8080/groot, and render "Hello, hoover".

📋 **Your version of the Python**

3.8

📋 **Your version of the aiohttp/yarl/multidict distributions**

aiohttp 3.7.2

### PR DESCRIPTION
**Pull request by danking**
## What do these changes do?

I think this fixes https://github.com/aio-libs/aiohttp/issues/5250.

A static resource like:
```
routes.static('/foo', '/www/foo')
```
currently matches `/foobar` but returns 404. This PR ensures that aforementioned static route *does not match* `/foobar`

## Are there changes in behavior for the user?

If a user expected `/foobar` to load the file `/www/foo/ar`, this PR will violate their expectation.

## Related issue number

Fixes #5250

## Checklist

I'm not aware of any docs that need to change, but I didn't tick that box because I am not certain.

- [X] I think the code is well written
- [X] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [X] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [X] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by Dreamsorcerer** [comment]
The fact that `/foobar` is translated to `/ar` seems surprising, so I'd be inclined to say that it's unlikely that people are depending on this behaviour. Though it could be possible that someone uses paths like `/static-filename`...

I wonder if the `prefix` argument should be renamed to `path` though, to avoid confusion if we do change it...

**Comment by danking** [comment]
OK, I’m fairly certain this is ready for review. Again, I don’t think the docs need to change because I think of this as a bug fix, but I’m not sure.

**Comment by Dreamsorcerer** [comment]
@webknjaz Thoughts on this? Seems to resolve some surprising behaviour.
But, looks like it also adds a little overhead, which I assume is running on every request.

Maybe a more efficient solution would be to ensure that that static resource is always at the end of the list, so it's the last one checked? This sounds reasonable to me, as a web server should be handling static assets on a production system where performance matters, therefore the static resource would never be useful.

File: CHANGES/5250.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+StaticResource prefixes no longer match URLs with a non-folder prefix. For example `routes.static('/foo', '/foo')` no longer matches the URL `/foobar`. Previously, this would attempt to load the file `/foo/ar`.
```
Comment (webknjaz) [inline CHANGES/5250.bugfix]: This note should use RST, not Markdown: please use double backticks.

File: tests/test_web_urldispatcher.py [hunk 2]
```diff
@@ -458,3 +458,18 @@ async def test_static_absolute_url(aiohttp_client: Any, tmp_path: Any) -> None:
...
+
+
+async def test_for_issue_5250(aiohttp_client: Any, tmp_path: Any) -> None:
+    app = web.Application()
+    app.router.add_static("/foo", tmp_path)
+
+    async def get_foobar(request):
+        return web.Response(body="success!")
+
+    app.router.add_get("/foobar", get_foobar)
```
Comment (webknjaz) [inline tests/test_web_urldispatcher.py:471]: I feel like there should be more tested cases and maybe some corner cases identified.

**Comment by webknjaz** [comment]
@Dreamsorcerer I'm not sure about that. Making the priority assumptions should probably be on the user. Besides, this would be a breaking change for people who used to rely on the order.

File: CHANGES/5250.bugfix (hunk 1 above)
Comment (asvetlov) [inline CHANGES/5250.bugfix]:
```suggestion
StaticResource prefixes no longer match URLs with a non-folder prefix. For example ``routes.static('/foo', '/foo')`` no longer matches the URL ``/foobar``. Previously, this would attempt to load the file ``/foo/ar``.
```

**Review by asvetlov — APPROVED** [review]
